### DISCUSSION TITLE
Get class by static field

### DISCUSSION THREAD

**Discussion by dykov** [pmd/pmd#3705] [Q&A] [answered]
Hi,

Is there any way to get class by its field?
For example, I need to check if the `org.apache.commons.codec.CharEncoding` class is being used:
```java
import org.apache.commons.codec.CharEncoding;
//...
Charset cs = Charset.forName(CharEncoding.UTF_8);
``` 
So, can I use something better than checking every static field of the class?
```java
//Name[
    pmd-java:typeIs('org.apache.commons.codec.CharEncoding.ISO_8859_1')
    or
    pmd-java:typeIs('org.apache.commons.codec.CharEncoding.UTF_8')
    (: or ... :)
]

Thank you

**Comment by jsotuyod** [comment, marked as answer]
First I have to acknowledge that undocumented / unintended usage of `typeIs`. Those static fields are actually Strings, but since PMD fails to resolve them to types, it assumes an incomplete auxclasspath and falls back to name matching, which actually works. Definitely not the intention, and will still produce config errors due to "incomplete auxclasspath".

Under PMD 6 you could do something along the lines (not tested):

```ruby
//ImportDeclaration/Name[@Image = 'org.apache.commons.codec.CharEncoding'] (: search for imports … :)
|
//Name[starts-with(@Image, 'org.apache.commons.codec.CharEncoding')] (: … or FQN usages:)
```

**Reply by dykov** [reply]
@jsotuyod , thanks for your advice. 
I have used a solution like this before, but faced the problem of suppressing for the import declaration.

**Reply by jsotuyod** [reply]
If you didn't want to flag the imports, and only the usages, I can't think of a way to do that without XPath 3's `fn:for-each` (which won't be supported until PMD 7 lands)

The best I can come up with is
```ruby
(: import + usage :)
//Name[//ImportDeclaration/Name[starts-with(@Image, 'org.apache.commons.codec.CharEncoding')]][starts-with(@Image, 'CharEncoding')]
|
(: FQN usage :)
//Name[not(parent::ImportDeclaration)][starts-with(@Image, 'org.apache.commons.codec.CharEncoding')]
|
(: static import + usage :)
//Name[not(parent::ImportDeclaration)][
  (: bind the context node to a variable :)
  some $name in (.) satisfies 
  //ImportDeclaration[@Static = true()]
    /Name[starts-with(@Image, 'org.apache.commons.codec.CharEncoding')]
    [ends-with(@Image, $name/@Image)]
```

Unfortunately, this would fail miserably with star static imports such as:

```java
import static org.apache.commons.codec.CharEncoding.*;

class Foo {
  Charset cs = Charset.forName(ISO_8859_1); // not flagged
  Charset cs2 = Charset.forName(UTF_8); // not flagged
}
```

without reflection these are impossible to get right.

Flagging the imports, or listing all constants in your hacky approach seems to be the only reliable options right now.

**Reply by oowekyala** [reply]
I've not tested it but doesn't this work and consider all imports?

```ruby
//Name[not(parent::ImportDeclaration)][
  (: bind the context node to a variable :)
  some $name in (.) satisfies 
  //ImportDeclaration[@Static = true()]
    /Name[starts-with(@Image, 'org.apache.commons.codec.CharEncoding')]
    [ends-with(@Image, $name/@Image)]
```

**Reply by jsotuyod** [reply]
yes, I believe it should. Nice thinking there, I was convinced `some … in … satisfies` was XPath 3 for some reason.
We would still miss star imports, but all static imports would work that way. I'll edit my response accordingly.

**Reply by dykov** [reply]
Thanks a lot to you all :)
